components/proxy/src/main/kotlin/com/hotels/styx/CommonTags.kt

```diff
@@ -0,0 +1,122 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx
+
+/**
+ * A common tag representation.
+ *
+ * Enforces a common tag format and provides a set of higher order functionality
+ * that operate on provided `name`, `encode` and `decode` properties.
+ *
+ * A tag string is a name-value pair separated by equal ('=') sign:
+ *
+ *     tag-string = name "=" value-part
+ *
+ * `encode` and `decode` are functions to encode a T to a value-part string,
+ * and to decode the value-part string back to T.
+ *
+ * @property name a tag name
+ * @property encode a function that encodes a value of T as a tag value string
+ * @property decode a function that decodes a tag value string as a T
+ */
+sealed class CommonValueTag<T>(
+        val name: String,
+        val encode: (T) -> String?,
+        val decode: (String) -> T?) {
```

Comment:
Is it worth providing defaults for encode and decode as `{ it }` ?

Reply:
Is this possible without knowing the type T?